File: Cpp/CXR_classify/mainwindow.cpp
#include "mainwindow.h"

MainWindow::MainWindow() : QMainWindow()
{
    configHandler = new ConfigHandler("./config.ini");

    configureLogging();
    logger->info("Constructing Main app");

    dbHandler = new DatabaseHandler(configHandler);

    setWindowTitle("CXR Classifier Training Walkthrough");

    fillWindow();
    initGuiState();
    show();

    logger->info("Done constructing Main app");
}

MainWindow::~MainWindow()
{
    delete configHandler;
}

void MainWindow::fillWindow()
{
    // Create widget for the dashboard
    QWidget * dashboardWidget = new QWidget();

    QLabel * msgBox = new QLabel("Welcome to the CXR Classification Application");
    QProgressBar * proBar = new QProgressBar;

    msgBox->setObjectName("msgBox");
    proBar->setObjectName("proBar");

    QGridLayout * dashboardLayout = new QGridLayout();
    dashboardLayout->addWidget(msgBox, 1, 0, 1, 3);
    dashboardLayout->addWidget(proBar, 2, 0, 1, 3);

    dashboardWidget->setLayout(dashboardLayout);

    // Create widget for the stage buttons
    QWidget * stagesWidget = new QWidget();

    QPushButton * downloadBtn = new QPushButton("Download");
    QPushButton * unpackBtn = new QPushButton("Unpack");
    QPushButton * storeBtn = new QPushButton("Store Metadata");
    QPushButton * featureBtn = new QPushButton("Calculate Features");
    QPushButton * labelBtn = new QPushButton("Label Images");
    QPushButton * trainBtn = new QPushButton("Train Classifier");

    downloadBtn->setObjectName("downloadBtn");
    unpackBtn->setObjectName("unpackBtn");
    storeBtn->setObjectName("storeBtn");
    featureBtn->setObjectName("featureBtn");
    labelBtn->setObjectName("labelBtn");
    trainBtn->setObjectName("trainBtn");

    QGridLayout * stagesLayout = new QGridLayout();
    stagesLayout->addWidget(downloadBtn, 1, 0);
    stagesLayout->addWidget(unpackBtn, 1, 1);
    stagesLayout->addWidget(storeBtn, 1, 2);
    stagesLayout->addWidget(featureBtn, 2, 0);
    stagesLayout->addWidget(labelBtn, 2, 1);
    stagesLayout->addWidget(trainBtn, 2, 2);

    stagesWidget->setLayout(stagesLayout);

    // Create widget for the labeler
    QWidget * labelerWidget = new QWidget();

    QLabel * image = new QLabel();
    image->setAlignment(Qt::AlignCenter);
    QPushButton * frontalBtn = new QPushButton("Frontal");
    QPushButton * lateralBtn = new QPushButton("Lateral");

    image->setObjectName("image");
    frontalBtn->setObjectName("frontalBtn");
    lateralBtn->setObjectName("lateralBtn");

    QGridLayout * labelLayout = new QGridLayout();
    labelLayout->addWidget(image, 1, 0, 1, 2);
    labelLayout->addWidget(frontalBtn, 2, 0);
    labelLayout->addWidget(lateralBtn, 2, 1);

    labelerWidget->setLayout(labelLayout);

    // Set up widget stack
    widgetStack = new QStackedWidget();
    widgetStack->addWidget(stagesWidget);
    widgetStack->addWidget(labelerWidget);

    // Full stack
    mainWidget = new QWidget(this);
    QVBoxLayout * mainLayout = new QVBoxLayout();
    mainLayout->addWidget(dashboardWidget);
    mainLayout->addWidget(widgetStack);
    mainWidget->setLayout(mainLayout);

    setCentralWidget(mainWidget);
}

void MainWindow::initGuiState()
{
    setWindowIcon(QIcon("../../miscellaneous/icon.jpg"));

    std::string folderRelPath = "./" + configHandler->getDatasetName();

    if (dbHandler->tableExists(configHandler->getTableName("label"))) {
        trainStageUi();
    } else if (dbHandler->tableExists(configHandler->getTableName("features"))) {
        labelStageUi();
    } else if (dbHandler->tableExists(configHandler->getTableName("metadata"))) {
        calcFeatStageUi();
    } else if (std::filesystem::exists(folderRelPath)) {
        storeStageUi();
    } else if (std::filesystem::exists(configHandler->getTgzFilename())) {
        unpackStageUi();
    } else {
        downloadStageUi();
    }
}

void MainWindow::downloadStageUi()
{
    logger->info("Window initializing in Download phase");
    currentStage = new DownloadStage(configHandler);

    disableAllStageButtons();
    enableStageButton(0);

    connect(mainWidget->findChild<QPushButton *>("downloadBtn"), SIGNAL (clicked()), static_cast<DownloadStage *>(currentStage), SLOT (download()));
    connectToDashboard(static_cast<DownloadStage *>(currentStage)->downloader);
    connect(static_cast<DownloadStage *>(currentStage)->downloader, SIGNAL (finished()), this, SLOT(clearCurrentStage()));
    connect(static_cast<DownloadStage *>(currentStage)->downloader, SIGNAL (finished()), this, SLOT(unpackStageUi()));
    logger->info("***Download phase initialized***");
}

void MainWindow::unpackStageUi()
{
    logger->info("Window initializing in Unpack phase");
    currentStage = new UnpackStage(configHandler);

    disableAllStageButtons();
    enableStageButton(1);

    connect(mainWidget->findChild<QPushButton *>("unpackBtn"), SIGNAL (clicked()), static_cast<UnpackStage *>(currentStage), SLOT (unpack()));
    connectToDashboard(static_cast<UnpackStage *>(currentStage)->unpacker);
    connect(static_cast<DownloadStage *>(currentStage)->downloader, SIGNAL (finished()), this, SLOT(clearCurrentStage()));
    connect(static_cast<UnpackStage *>(currentStage)->unpacker, SIGNAL (finished()), this, SLOT(storeStageUi()));
    logger->info("***Unpack phase initialized***");
}

void MainWindow::storeStageUi()
{
    logger->info("Window initializing in Store phase");
    currentStage = new StoreStage(configHandler, dbHandler);

    disableAllStageButtons();
    enableStageButton(2);

    connect(mainWidget->findChild<QPushButton *>("storeBtn"), SIGNAL (clicked()), static_cast<StoreStage *>(currentStage), SLOT (store()));
    connectToDashboard(static_cast<StoreStage *>(currentStage)->storer);
    connect(static_cast<DownloadStage *>(currentStage)->downloader, SIGNAL (finished()), this, SLOT(clearCurrentStage()));
    connect(static_cast<StoreStage *>(currentStage)->storer, SIGNAL (finished()), this, SLOT(calcFeatStageUi()));
    logger->info("***Store phase initialized***");
}

void MainWindow::calcFeatStageUi()
{
    logger->info("Window initializing in Feature Calculation phase");
    currentStage = new FeatureCalculatorStage(configHandler, dbHandler);

    disableAllStageButtons();
    enableStageButton(3);

    connect(mainWidget->findChild<QPushButton *>("featureBtn"), SIGNAL (clicked()), static_cast<FeatureCalculatorStage *>(currentStage), SLOT (calculateFeatures()));
    connectToDashboard(static_cast<FeatureCalculatorStage *>(currentStage)->featureCalculator);
    connect(static_cast<DownloadStage *>(currentStage)->downloader, SIGNAL (finished()), this, SLOT(clearCurrentStage()));
    connect(static_cast<FeatureCalculatorStage *>(currentStage)->featureCalculator, SIGNAL (finished()), this, SLOT(labelStageUi()));
    logger->info("***Feature Calculation phase initialized***");
}

void MainWindow::labelStageUi()
{
    logger->info("Window initializing in Labeling phase");
    currentStage = new LabelStage(configHandler, dbHandler);

    disableAllStageButtons();
    enableStageButton(4);

    connect(mainWidget->findChild<QPushButton *>("labelBtn"), SIGNAL (clicked()), static_cast<LabelStage *>(currentStage), SLOT (label()));

    if (configHandler->getDatasetType() == "subset") {
        connectToDashboard(static_cast<LabelStage *>(currentStage)->labeler);

        connect(widgetStack->findChild<QPushButton *>("frontalBtn"), SIGNAL (clicked()), static_cast<LabelStage *>(currentStage)->labeler, SLOT (frontal()));
        connect(widgetStack->findChild<QPushButton *>("lateralBtn"), SIGNAL (clicked()), static_cast<LabelStage *>(currentStage)->labeler, SLOT (lateral()));

        connect(mainWidget->findChild<QPushButton *>("labelBtn"), SIGNAL (clicked()), this, SLOT (secondPage()));
        connect(static_cast<LabelStage *>(currentStage)->labeler, SIGNAL (finished()), this, SLOT (firstPage()));
        connect(static_cast<DownloadStage *>(currentStage)->downloader, SIGNAL (finished()), this, SLOT(clearCurrentStage()));
        connect(static_cast<LabelStage *>(currentStage)->labeler, SIGNAL (finished()), this, SLOT (trainStageUi()));
    } else {
        connectToDashboard(static_cast<LabelStage *>(currentStage)->labeler);
        connect(static_cast<DownloadStage *>(currentStage)->downloader, SIGNAL (finished()), this, SLOT(clearCurrentStage()));
        connect(static_cast<LabelStage *>(currentStage)->labeler, SIGNAL (finished()), this, SLOT (trainStageUi()));
    }
    logger->info("***Labeling phase initialized***");
}

void MainWindow::trainStageUi()
{
    logger->info("Window initializing in Training phase");
    currentStage = new TrainStage(configHandler, dbHandler);

    widgetStack->setFixedSize(widgetStack->currentWidget()->layout()->sizeHint());
    setFixedSize(centralWidget()->layout()->sizeHint());

    disableAllStageButtons();
    enableStageButton(5);

    connect(mainWidget->findChild<QPushButton *>("trainBtn"), SIGNAL (clicked()), static_cast<TrainStage *>(currentStage), SLOT (train()));
    connectToDashboard(static_cast<TrainStage *>(currentStage)->trainer);
    logger->info("***Training phase initialized***");
}

void MainWindow::clearCurrentStage()
{
    delete currentStage;
}

void MainWindow::firstPage()
{
    widgetStack->setCurrentIndex(0);
}

void MainWindow::secondPage()
{
    widgetStack->setCurrentIndex(1);
}

void MainWindow::connectToDashboard(Runnable * stage)
{
    connect(stage, SIGNAL (attemptUpdateProBarBounds(quint64, quint64)), this, SLOT (updateProBarBounds(quint64, quint64)));
    connect(stage, SIGNAL (attemptUpdateProBarValue(quint64)), this, SLOT (updateProBarValue(quint64)));
    connect(stage, SIGNAL (attemptUpdateText(QString)), this, SLOT (updateText(QString)));
    connect(stage, SIGNAL (attemptUpdateImage(QPixmap)), this, SLOT (updateImage(QPixmap)));
}

void MainWindow::disableAllStageButtons()
{
    uint8_t numOfButtons = sizeof(buttonsList)/sizeof(buttonsList[0]);
    for (uint8_t i = 0; i < numOfButtons; i++) {
        mainWidget->findChild<QPushButton *>(buttonsList[i])->setDisabled(true);
    }
}

void MainWindow::enableStageButton(quint64 stageIndex)
{
    mainWidget->findChild<QPushButton *>(buttonsList[stageIndex])->setDisabled(false);
}

void MainWindow::configureLogging()
{
    try {
        logger = spdlog::basic_logger_mt(loggerName, "CXR_Classification.log");
    }
    catch (const spdlog::spdlog_ex &ex) {
        std::cout << "Log init failed: " << ex.what() << std::endl;
    }

    std::string logLevel = configHandler->getLogLevel();
    if (logLevel == "debug") {
        logger->set_level(spdlog::level::debug);
    } else if (logLevel == "info") {
        logger->set_level(spdlog::level::info);
    }
}

void MainWindow::updateText(QString text)
{
    mainWidget->findChild<QLabel *>("msgBox")->setText(text);
}

void MainWindow::updateProBarBounds(quint64 proBarMin, quint64 proBarMax)
{
    mainWidget->findChild<QProgressBar *>("proBar")->setMinimum(proBarMin);
    mainWidget->findChild<QProgressBar *>("proBar")->setMaximum(proBarMax);
}

void MainWindow::updateProBarValue(quint64 value) {
    mainWidget->findChild<QProgressBar *>("proBar")->setValue(value);
}

void MainWindow::updateImage(QPixmap pixmap) {
    widgetStack->findChild<QLabel *>("image")->setPixmap(pixmap);
}

File: Python/DesktopApp/main_window.py
import logging
from PyQt5.QtCore import pyqtSlot, QThread, Qt
from PyQt5.QtGui import QPixmap, QImage
from PyQt5.QtWidgets import QMainWindow, QWidget, QProgressBar, QLabel, QPushButton, QStackedWidget, QGridLayout, QVBoxLayout
import pydicom as pdm
import cv2
import numpy as np

class MainWindow(QMainWindow):
    """Contains GUI code for the application."""
    def __init__(self, controller):
        logging.info('Constructing Main app')
        QMainWindow.__init__(self, windowTitle="CXR Classifier Training Walkthrough")
        self.controller = controller

        self.buttons_list = ["download_btn", "unpack_btn", "store_btn", "features_btn", "label_btn", "classify_btn"]

        self.fill_window()
        self.show()
        
        logging.info('Done constructing Main app')

    def fill_window(self):
        """Fills the window with buttons."""
        # Create widget for the dashboard
        dashboardWidget = QWidget()

        msg_box = QLabel('Welcome to the CXR Classification Application', objectName="msg_box")
        pro_bar = QProgressBar(objectName="pro_bar")

        dashboardLayout = QGridLayout()
        dashboardLayout.addWidget(msg_box, 1, 0, 1, 3)
        dashboardLayout.addWidget(pro_bar, 2, 0, 1, 3)

        dashboardWidget.setLayout(dashboardLayout)

        # Create widget for the stage buttons
        stagesWidget = QWidget()

        download_btn = QPushButton("Download", objectName="download_btn", clicked=self.controller.downloadStage.download)
        unpack_btn = QPushButton("Unpack", objectName="unpack_btn", clicked=self.controller.unpackStage.unpack)
        store_btn = QPushButton("Store Metadata", objectName="store_btn", clicked=self.controller.storeStage.store)
        features_btn = QPushButton("Calculate Features", objectName="features_btn", clicked=self.controller.featCalcStage.calculateFeatures)
        label_btn = QPushButton("Label Images", objectName="label_btn")
        classify_btn = QPushButton("Train Classifier", objectName="classify_btn", clicked=self.controller.trainStage.train)
        
        stagesLayout = QGridLayout()
        stagesLayout.addWidget(download_btn, 1, 0)
        stagesLayout.addWidget(unpack_btn, 1, 1)
        stagesLayout.addWidget(store_btn, 1, 2)
        stagesLayout.addWidget(features_btn, 2, 0)
        stagesLayout.addWidget(label_btn, 2, 1)
        stagesLayout.addWidget(classify_btn, 2, 2)

        stagesWidget.setLayout(stagesLayout)

        # Create widget for the labeler
        labelerWidget = QWidget()

        image = QLabel(objectName="image")
        image.setAlignment(Qt.AlignCenter)
        frontal_btn = QPushButton('Frontal', objectName="frontal_btn", clicked=self.controller.labelerStage.labeler.frontal)
        lateral_btn = QPushButton('Lateral', objectName="lateral_btn", clicked=self.controller.labelerStage.labeler.lateral)

        labelLayout = QGridLayout()
        labelLayout.addWidget(image, 1, 0, 1, 2)
        labelLayout.addWidget(frontal_btn, 2, 0)
        labelLayout.addWidget(lateral_btn, 2, 1)

        labelerWidget.setLayout(labelLayout)
        
        # Set up widget stack
        self.widgetStack = QStackedWidget()
        self.widgetStack.addWidget(stagesWidget)
        self.widgetStack.addWidget(labelerWidget)

        # Full stack
        self.mainWidget = QWidget()
        mainLayout = QVBoxLayout()
        mainLayout.addWidget(dashboardWidget)
        mainLayout.addWidget(self.widgetStack)
        self.mainWidget.setLayout(mainLayout)

        self.setCentralWidget(self.mainWidget)

    @pyqtSlot(int)
    def update_pro_bar_val(self, value):
        self.centralWidget().findChild(QProgressBar, "pro_bar").setValue(value)

    @pyqtSlot(int, int)
    def update_pro_bar_bounds(self, proBarMin, proBarMax):
        self.centralWidget().findChild(QProgressBar, "pro_bar").setMinimum(proBarMin)
        self.centralWidget().findChild(QProgressBar, "pro_bar").setMaximum(proBarMax)

    @pyqtSlot(str)
    def update_text(self, text):
        self.centralWidget().findChild(QLabel, "msg_box").setText(text)

    @pyqtSlot(object)
    def updateImage(self, record):
        image = pdm.dcmread(record['file_path']).pixel_array
        bits_stored = record['bits_stored']
        pixmap = self.arr_into_pixmap(image, bits_stored)
        self.widgetStack.currentWidget().findChild(QLabel, "image").setPixmap(pixmap)

    @pyqtSlot()
    def downloadStageUi(self):
        logging.info('Window initializing in Download phase')
        self.disableAllStageButtons()
        self.enableStageButton(0)

        self.connectToDashboard(self.controller.downloadStage.downloader.signals)

        self.controller.downloadStage.downloader.signals.finished.connect(self.unpackStageUi)
        logging.info('***Download phase initialized***')

    @pyqtSlot()
    def unpackStageUi(self):
        logging.info('Window initializing in Unpack phase')
        self.disableAllStageButtons()
        self.enableStageButton(1)

        self.connectToDashboard(self.controller.unpackStage.unpackUpdater.signals)

        self.controller.unpackStage.unpackUpdater.signals.finished.connect(self.storeStageUi)
        logging.info('***Unpack phase initialized***')
        
    @pyqtSlot()
    def storeStageUi(self):
        logging.info('Window initializing in Store phase')
        self.disableAllStageButtons()
        self.enableStageButton(2)

        self.connectToDashboard(self.controller.storeStage.storeUpdater.signals)

        self.controller.storeStage.storeUpdater.signals.finished.connect(self.calcFeatStageUi)
        logging.info('***Store phase initialized***')

    @pyqtSlot()
    def calcFeatStageUi(self):
        logging.info('Window initializing in Feature Calculation phase')
        self.disableAllStageButtons()
        self.enableStageButton(3)

        self.connectToDashboard(self.controller.featCalcStage.featureCalculator.signals)

        self.controller.featCalcStage.featureCalculator.signals.finished.connect(self.labelStageUi)
        logging.info('***Feature Calculation phase initialized***')

    @pyqtSlot()
    def labelStageUi(self):
        logging.info('Window initializing in Labeling phase')
        self.disableAllStageButtons()
        self.enableStageButton(4)

        if self.controller.configHandler.getDatasetType() == 'subset':
            self.centralWidget().findChild(QPushButton, "label_btn").clicked.connect(lambda: self.widgetStack.setCurrentIndex(1))
            self.centralWidget().findChild(QPushButton, "label_btn").clicked.connect(self.controller.labelerStage.label)

            self.controller.labelerStage.labeler.signals.attemptUpdateImage.connect(self.updateImage)

            self.connectToDashboard(self.controller.labelerStage.labeler.signals)
            self.controller.labelerStage.labeler.signals.finished.connect(lambda: self.widgetStack.setCurrentIndex(0))
            self.controller.labelerStage.labeler.signals.finished.connect(self.trainStageUi)

        elif self.controller.configHandler.getDatasetType() == 'full_set':
            self.centralWidget().findChild(QPushButton, "label_btn").clicked.connect(self.controller.labelImporterStage.importLabels)
            self.connectToDashboard(self.controller.labelImporterStage.labelImporter.signals)
            self.controller.labelImporterStage.labelImporter.signals.finished.connect(self.trainStageUi)
        else:
            raise ValueError('Value must be one of the keys in SOURCE_URL')
        logging.info('***Labeling phase initialized***')

    @pyqtSlot()
    def trainStageUi(self):
        logging.info('Window initializing in Training phase')
        self.widgetStack.setFixedSize(self.widgetStack.currentWidget().layout().sizeHint())
        self.setFixedSize(self.centralWidget().layout().sizeHint())

        self.disableAllStageButtons()
        self.enableStageButton(5)

        self.connectToDashboard(self.controller.trainStage.trainer.signals)
        logging.info('***Training phase initialized***')

    def connectToDashboard(self, signals):
        signals.attemptUpdateProBarBounds.connect(self.update_pro_bar_bounds)
        signals.attemptUpdateProBarValue.connect(self.update_pro_bar_val)
        signals.attemptUpdateText.connect(self.update_text)

    def disableAllStageButtons(self):
        for button in self.buttons_list:
            self.centralWidget().findChild(QPushButton, button).setDisabled(True)

    def enableStageButton(self, stageIndex):
        self.centralWidget().findChild(QPushButton, self.buttons_list[stageIndex]).setDisabled(False)

    def arr_into_pixmap(self, image, bits_stored):
        """Convert the image array into a QPixmap for display."""
        # Scale the pixel intensity to uint8
        highest_possible_intensity = (np.power(2, bits_stored) - 1)
        image = (image/highest_possible_intensity * 255).astype(np.uint8)

        image = cv2.resize(image, (300,300), interpolation=cv2.INTER_AREA)

        height, width = image.shape
        bytes_per_line = width
        q_image = QImage(image, width, height, bytes_per_line, QImage.Format_Grayscale8)
        pixmap = QPixmap.fromImage(q_image)

        return pixmap
